server/src/repositories/person.repository.ts

```diff
@@ -83,14 +84,32 @@ export class PersonRepository implements IPersonRepository {
   }
 
   @GenerateSql({ params: [{ take: 10, skip: 10 }, DummyValue.UUID] })
-  getAllForUser(pagination: PaginationOptions, userId: string, options?: PersonSearchOptions): Paginated<PersonEntity> {
+  async getAllForUser(
+    pagination: PaginationOptions,
+    userId: string,
+    options?: PersonSearchOptions,
+  ): Paginated<PersonEntity> {
     const queryBuilder = this.personRepository
       .createQueryBuilder('person')
       .innerJoin('person.faces', 'face')
       .where('person.ownerId = :userId', { userId })
       .innerJoin('face.asset', 'asset')
       .andWhere('asset.isArchived = false')
-      .orderBy('person.isHidden', 'ASC')
+      .orderBy('person.isHidden', 'ASC');
+    if (options?.closestFaceAssetId) {
+      const face = await this.faceSearchRepository.findOne({
+        where: { faceId: options.closestFaceAssetId },
+      });
+      if (!face?.embedding) {
+        throw new Error('Face does not exist');
+      }
+      queryBuilder
+        .leftJoin('face.faceSearch', 'search')
+        .addSelect('AVG(search.embedding <=> :embedding)', 'distance')
+        .addOrderBy('distance')
```

Comment:
This is a brute-force comparison across all face embeddings that doesn't use any index. Restructure the query to leverage the face index. This means adding a limit and (re)moving the averaging.

Reply:
I changed it to only comparing the thumbnails for every person instead of all the faces. Not sure if that is what you meant though.

Reply:
This is definitely better, but I don't think this would use an index either because there's no limit, and maybe also because of the other order and grouping conditions.

This could be okay depending on the query plan. Could you run this query with EXPLAIN (ANALYZE, BUFFERS) and share the output? Also try adding a limit of 20 and share that version of the plan.

Reply:
Here is without a limit (I ran this on my personal instance)
```
Sort  (cost=38108.28..38112.58 rows=1722 width=214) (actual time=150.680..150.788 rows=4748 loops=1)
   Sort Key: person."isHidden", (((SubPlan 1) <=> $1)), ((NULLIF((person.name)::text, ''::text) IS NULL)), (count(face."assetId")) DESC, (NULLIF((person.name)::text, ''::text)), person."createdAt"
   Sort Method: quicksort  Memory: 1454kB
   Buffers: shared hit=70805, temp read=996 written=996
   InitPlan 2 (returns $1)
     ->  Index Scan using face_search_pkey on face_search face_search_1  (cost=0.42..8.44 rows=1 width=18) (actual time=0.009..0.010 rows=1 loops=1)
           Index Cond: ("faceId" = 'ce08488c-0d6d-4275-b72d-454ed585fb7b'::uuid)
           Buffers: shared hit=4
   ->  HashAggregate  (cost=23393.05..38007.28 rows=1722 width=214) (actual time=116.271..146.949 rows=4748 loops=1)
         Group Key: person.id
         Filter: (((person.name)::text <> ''::text) OR (count(face."assetId") >= 1))
         Batches: 1  Memory Usage: 1745kB
         Buffers: shared hit=70805, temp read=996 written=996
         ->  Hash Join  (cost=15991.02..22759.94 rows=126622 width=185) (actual time=63.838..106.708 rows=106214 loops=1)
               Hash Cond: (face."personId" = person.id)
               Buffers: shared hit=14282, temp read=996 written=996
               ->  Hash Join  (cost=15725.28..22140.30 rows=134727 width=32) (actual time=58.676..89.775 rows=133470 loops=1)
                     Hash Cond: (face."assetId" = asset.id)
                     Buffers: shared hit=14140, temp read=996 written=996
                     ->  Seq Scan on asset_faces face  (cost=0.00..3525.76 rows=136476 width=32) (actual time=0.014..7.429 rows=134602 loops=1)
                           Buffers: shared hit=2161
                     ->  Hash  (cost=13358.28..13358.28 rows=136160 width=16) (actual time=57.580..57.580 rows=122348 loops=1)
                           Buckets: 131072  Batches: 4  Memory Usage: 2459kB
                           Buffers: shared hit=11979, temp written=403
                           ->  Seq Scan on assets asset  (cost=0.00..13358.28 rows=136160 width=16) (actual time=0.022..35.342 rows=122348 loops=1)
                                 Filter: (("deletedAt" IS NULL) AND (NOT "isArchived"))
                                 Rows Removed by Filter: 1524
                                 Buffers: shared hit=11979
               ->  Hash  (cost=205.79..205.79 rows=4796 width=169) (actual time=5.123..5.123 rows=4748 loops=1)
                     Buckets: 8192  Batches: 1  Memory Usage: 1019kB
                     Buffers: shared hit=142
                     ->  Seq Scan on person  (cost=0.00..205.79 rows=4796 width=169) (actual time=0.016..2.049 rows=4748 loops=1)
                           Filter: ((NOT "isHidden") AND ("ownerId" = '08bfaa34-c942-4721-91c9-f72b04c48d59'::uuid))
                           Rows Removed by Filter: 314
                           Buffers: shared hit=142
         SubPlan 1
           ->  Index Scan using face_search_pkey on face_search  (cost=0.42..8.44 rows=1 width=18) (actual time=0.002..0.002 rows=1 loops=4748)
                 Index Cond: ("faceId" = person."faceAssetId")
                 Buffers: shared hit=18832
 Planning:
   Buffers: shared hit=52
 Planning Time: 1.124 ms
 Execution Time: 151.388 ms
```
Here is with a limit
```
Sort  (cost=38108.28..38112.58 rows=1722 width=214) (actual time=116.916..117.023 rows=4748 loops=1)
   Sort Key: person."isHidden", (((SubPlan 1) <=> $1)), ((NULLIF((person.name)::text, ''::text) IS NULL)), (count(face."assetId")) DESC, (NULLIF((person.name)::text, ''::text)), person.
"createdAt"
   Sort Method: quicksort  Memory: 1454kB
   Buffers: shared hit=70782, temp read=996 written=996
   InitPlan 2 (returns $1)
     ->  Limit  (cost=0.42..8.44 rows=1 width=18) (actual time=0.004..0.005 rows=1 loops=1)
           Buffers: shared hit=4
           ->  Index Scan using face_search_pkey on face_search face_search_1  (cost=0.42..8.44 rows=1 width=18) (actual time=0.004..0.004 rows=1 loops=1)
                 Index Cond: ("faceId" = 'ce08488c-0d6d-4275-b72d-454ed585fb7b'::uuid)
                 Buffers: shared hit=4
   ->  HashAggregate  (cost=23393.05..38007.28 rows=1722 width=214) (actual time=88.435..113.227 rows=4748 loops=1)
         Group Key: person.id
         Filter: (((person.name)::text <> ''::text) OR (count(face."assetId") >= 1))
         Batches: 1  Memory Usage: 1745kB
         Buffers: shared hit=70782, temp read=996 written=996
         ->  Hash Join  (cost=15991.02..22759.94 rows=126622 width=185) (actual time=36.697..78.875 rows=106214 loops=1)
               Hash Cond: (face."personId" = person.id)
               Buffers: shared hit=14282, temp read=996 written=996
               ->  Hash Join  (cost=15725.28..22140.30 rows=134727 width=32) (actual time=32.409..62.688 rows=133470 loops=1)
                     Hash Cond: (face."assetId" = asset.id)
                     Buffers: shared hit=14140, temp read=996 written=996
                     ->  Seq Scan on asset_faces face  (cost=0.00..3525.76 rows=136476 width=32) (actual time=0.009..7.105 rows=134602 loops=1)
                           Buffers: shared hit=2161
                     ->  Hash  (cost=13358.28..13358.28 rows=136160 width=16) (actual time=32.201..32.202 rows=122348 loops=1)
                           Buckets: 131072  Batches: 4  Memory Usage: 2459kB
                           Buffers: shared hit=11979, temp written=403
                           ->  Seq Scan on assets asset  (cost=0.00..13358.28 rows=136160 width=16) (actual time=0.010..21.714 rows=122348 loops=1)
                                 Filter: (("deletedAt" IS NULL) AND (NOT "isArchived"))
                                 Rows Removed by Filter: 1524
                                 Buffers: shared hit=11979
               ->  Hash  (cost=205.79..205.79 rows=4796 width=169) (actual time=4.260..4.260 rows=4748 loops=1)
                     Buckets: 8192  Batches: 1  Memory Usage: 1019kB
                     Buffers: shared hit=142
                     ->  Seq Scan on person  (cost=0.00..205.79 rows=4796 width=169) (actual time=0.016..1.949 rows=4748 loops=1)
                           Filter: ((NOT "isHidden") AND ("ownerId" = '08bfaa34-c942-4721-91c9-f72b04c48d59'::uuid))
                           Rows Removed by Filter: 314
                           Buffers: shared hit=142
         SubPlan 1
           ->  Limit  (cost=0.42..8.44 rows=1 width=18) (actual time=0.001..0.001 rows=1 loops=4748)
                 Buffers: shared hit=18832
                 ->  Index Scan using face_search_pkey on face_search  (cost=0.42..8.44 rows=1 width=18) (actual time=0.001..0.001 rows=1 loops=4748)
                       Index Cond: ("faceId" = person."faceAssetId")
                       Buffers: shared hit=18832
 Planning:
   Buffers: shared hit=52
 Planning Time: 1.208 ms
 Execution Time: 117.367 ms
```
Just for clarification the limit was added to the selection queries
```js
const innerQueryBuilder = this.faceSearchRepository
        .createQueryBuilder('face_search')
        .select('embedding', 'embedding')
        .where('"face_search"."faceId" = "person"."faceAssetId"').limit(1);
const faceSelectQueryBuilder = this.faceSearchRepository
        .createQueryBuilder('face_search')
        .select('embedding', 'embedding')
        .where('"face_search"."faceId" = :faceId', { faceId: options.closestFaceAssetId }).limit(1);
```

Reply:
Nice. Can you also change the query so it doesn't order by the other conditions if the similarity search is used? They're useless in this case so no need for that overhead.

Reply:
The limits you added aren't necessary. I meant having a limit for the whole query, but I forgot this is already a paginated endpoint.

Reply:
I forgot to respond this morning, but I did push the change for only using the similarity orderBy for when similarity search is used.